proposals/simd-0170-builtin-instruction-cost-and-budget.md

```diff
@@ -0,0 +1,120 @@
+---
+simd: '0170'
+title: Reserve minimal CUs for builtins
+authors:
+  - Tao Zhu (Anza)
+category: Standard
+type: Core
+status: Review
+created: 2024-08-26
+feature: https://github.com/anza-xyz/agave/issues/2562
+supersedes: 
+superseded-by:
+extends:
+---
+
+## Summary
+
+When transactions do not request a specific compute unit limit, the Solana
+runtime conservatively allocates 200,000 compute units for each tx-level
+instruction (excluding compute budget instructions) in the transaction's
+compute budget. Since builtin program instructions consume far less than 200k
+compute units, the runtime will be modified to only allocate 3,000 compute units
+in the transaction compute budget for each of these instructions.
+
+## Motivation
+
+When allocating the transaction compute budget for transactions which don't
+specify a specific compute unit limit, the Solana runtime over-allocates the
+number of compute units needed for executing builtin programs which have a
+small (mostly) static execution compute unit cost. The runtime will only release
+the over-allocated compute units from the block cost tracker when transaction
+execution finishes and the actual consume compute units is known.
+
+Over-allocating CUs for builtin program instructions reduces block density,
+lowers network throughput, and can degrade block producer performance by causing
+repeated transaction retries (due to repeatedly over allocating cu's and then
+releasing them after execution). Avoiding excessive CU allocation is critical
+for maximizing block efficiency and minimizing network delays.
+
+## New Terminology
+
+None
+
+## Detailed Design
+
+Establish a static maximum allocation of 3,000 CU's for builtin programs denoted
+as `MAX_BUILTIN_ALLOCATION_COMPUTE_UNIT_LIMIT`, which is larger than all actual
+compute unit costs of builtin programs as well as the potential Cross-Program
+Invocations (CPIs) they do. This uniform limit is applied to each builtin
+program instruction and is used to configure both CU meter allocations and block
+producer CU limits reservation for all builtin instructions.
+
+The static list of builtin program id's that will have 3000 compute units
+allocated are:
```

Comment:
Can you add a line here that also states the builtin must be owned by the native loader for this CU allocation to apply? Avoids any footguns when they move to BPF.

Reply:
The list can still be static, we just also want that explicit requirement.

Reply:
added in 0e6e0caa2d35a915ae2455dadf66a4ad8c1e8ae6

Reply:
I think we should update this list of builtin program id's when each of the builtin-to-bpf feature gates are activated. We're not going to check the owner of the builtin program each time during cost calculation.

Reply:
Works for me, as long as an ID can be popped out on feature activation.

Reply:
Yup, @tao-stones please make sure @buffalojoec gets added as a reviewer on your impl pr

Reply:
yep, will do